File: sources/csharp/05_alloc_churn/05_alloc_churn.cs
// task 05 alloc_churn — expected output: 1274991808
// build: coreclr: dotnet build -c Release | nativeaot: dotnet publish -c Release -p:PublishAot=true | mono: mcs -optimize+ 05_alloc_churn.cs    run: coreclr: dotnet run -c Release (or bin/Release/net8.0/05_alloc_churn.exe) | nativeaot: bin/Release/net8.0/publish/05_alloc_churn.exe | mono: mono 05_alloc_churn.exe
using System;

class Program
{
    static void Main()
    {
        long total = 0;
        byte[][] slots = new byte[256][];
        for (int i = 0; i < 10000000; i++)
        {
            byte[] buf = new byte[64];
            buf[0] = (byte)(i % 256);
            total += buf[0];
            // Storing the buffer keeps it reachable; the array it replaces becomes garbage.
            slots[i % 256] = buf;
        }
        Console.WriteLine(total);
    }
}

File: sources/vbnet/05_alloc_churn/05_alloc_churn.vb
' task 05 alloc_churn — expected output: 1274991808
' build: dotnet build -c Release 05_alloc_churn.vbproj    run: dotnet bin/Release/net8.0/05_alloc_churn.dll

Imports System

Module Program
    Sub Main()
        Dim total As Long = 0
        Dim slots(255)() As Byte

        For i As Long = 0 To 9999999
            Dim buf(63) As Byte
            buf(0) = CByte(i Mod 256)
            total += buf(0)
            slots(CInt(i Mod 256)) = buf
        Next

        Console.WriteLine(total)
    End Sub
End Module
